tests/cmds/test_users.py

```diff
@@ -0,0 +1,88 @@
+import json
+
+import pytest
+from pandas.core.frame import DataFrame
+from py42.response import Py42Response
+from requests import Response
+
+from code42cli.cmds.users import _get_role_id
+from code42cli.cmds.users import _get_users_dataframe
+from code42cli.main import cli
+
+TEST_ROLE_RETURN_DATA = {"data": [{"roleName": "Customer Cloud Admin", "roleId": "12"}]}
+
+TEST_USERS_RESPONSE = {
+    "users": [
+        {
+            "userId": "1234",
+            "userUid": "997962681513153325",
+            "status": "Active",
+            "username": "test_username@code42.com",
+            "creationDate": "2021-03-12T20:07:40.898Z",
+            "modificationDate": "2021-03-12T20:07:40.938Z",
+        }
+    ]
+}
+
+
+def _create_py42_response(mocker, text):
+    response = mocker.MagicMock(spec=Response)
+    response.text = text
+    response._content_consumed = mocker.MagicMock()
+    response.status_code = 200
+    return Py42Response(response)
+
+
+def get_all_users_generator():
+    yield TEST_USERS_RESPONSE
+
+
+@pytest.fixture
+def get_available_roles_response(mocker):
+    return _create_py42_response(mocker, json.dumps(TEST_ROLE_RETURN_DATA))
+
+
+@pytest.fixture
+def get_all_users_success(cli_state):
+    cli_state.sdk.users.get_all.return_value = get_all_users_generator()
+
+
+@pytest.fixture
+def get_available_roles_success(cli_state, get_available_roles_response):
+    cli_state.sdk.users.get_available_roles.return_value = get_available_roles_response
+
+
+def test_list_outputs_appropriate_columns(runner, cli_state, get_all_users_success):
+    result = runner.invoke(cli, ["users", "list"], obj=cli_state)
+    assert "userId" in result.output
+    assert "userUid" in result.output
+    assert "status" in result.output
+    assert "username" in result.output
+    assert "creationDate" in result.output
+    assert "modificationDate" in result.output
+
+
+def test_get_role_id_returns_appropriate_role_id(
+    cli_state, get_available_roles_success
+):
+    result = _get_role_id(cli_state.sdk, "Customer Cloud Admin")
+    assert result == "12"
+
+
+def test_get_users_dataframe_returns_dataframe(cli_state, get_all_users_success):
+    result = _get_users_dataframe(
+        cli_state.sdk, org_uid=None, role_id=None, active=None
+    )
+    assert isinstance(result, DataFrame)
+
+
+def test_get_users_dataframe_calls_get_all_users_with_correct_parameters(
```

Comment:
We don't typically test the implementation details of commands (the protected methods).
Can we change these tests such that their entry point occurs from command execution?

Reply:
Updated!

Reply:
There are also a couple other tests (test_get_role_id_returns_appropriate_role_id and test_get_users_dataframe_returns_dataframe) that test the details of internal methods, should those be updated as well? I think it might be difficult to test some of those things from command execution as they aren't directly exposed in the result.

Reply:
Yes, that would be nice. 

Tests like `test_get_users_dataframe_returns_dataframe` are not needed. The fact that we use dataframes under the hood shouldn't matter.

For testing `role_ids` logic, the test would look something this 

```
def test_list_users_includes_expected_role_ids()
```

And then you would verify that the role IDs from the mock responses have shown up in the output

```
assert "12" in result.output
```

(Mayne make the role ID more specific, like `1230123` or something)


Additionally, I would really appreciate some tests around the logic that figures out `active`.

```
def test_list_users_when_given_active_uses_inactivate_uses_active_equals_false()

def test_list_users_when_given_active_and_inactive_raises_error()

def test_list_users_when_given_excluding_active_and_inactive_uses_active_equals_none()
```

^ These tests can just verify that py42 was called correctly if we are unable to verify these states via the output.

Reply:
I've added the additional tests for the active logic. I updated the role Id test as well, but roleId isn't actually included in the output, it's just passed as a parameter to the users.get_all() method that gets the users (where it's used as a filter). So I ended up just testing that the users.get_all() call has the appropriate parameter.